kernel/src/expressions/scalars.rs

```diff
@@ -13,6 +13,22 @@ pub struct StructData {
     values: Vec<Scalar>,
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArrayData {
+    tpe: ArrayType,
+    pub elements: Vec<Scalar>,
```

Comment:
Hmmm, I see your point, but I don't have an immediate idea for what might be best here. Off the top, engine data across the board instead of introducing another opaque time I think would be a better option (assume it's reasonable achievable). 

This code was written very much to suit delta-rs's partition filtering logic which rarely has more than a few values, I've only seen it get into the thousands a few times in the wild for a hive style partition filter. But more to your other point about subqueries and BI tools generating large constant lists, I have sadly seen those hit the max length of a java array and I have seen more than a few that are in the hundreds of thousands or millions. So your concern is well placed, but perhaps given more conversation about the representation would it be acceptable to table that discussion for a subsequent PR?

Reply:
Yeah, we don't need to solve it immediately, but definitely worth a tracking issue because it will bite us eventually. The advantage of going engine data route is then the engine can decide how to layout the data, whether to build a hash table, etc -- which seems very in line with kernel philosophy. I don't know how easily we could abstract that away without going the engine data route.

Reply:
Not to solution too much here, but I think this would make sense as an `EngineDataLayout` trait of sorts, where certain layout specific constants are filled in. This would be somewhat of a combination of `OffsetSizeTrait` `ArrowNativeType` and `ArrowPrimitiveType` in which the end implementation denotes layout specific information about it's engine data structure. This also has some interesting outside applications from arrays for example duckdb has a row_idx type for their row IDs. it's a uint64_t for them, but for another engine this may not be the case. I think we should reasonably strive to speak the same structure of the engine when dealing with it's data, yes?

Reply:
Thinking more about this situation -- we're choosing to model [NOT] IN as a binary expression, which requires us to introduce some new expression type that can represent the RHS. We could go one of two ways here:
1. Introduce a general array literal type, which is transparent to kernel and which [NOT] IN expressions happen to employ as their RHS.
2. Introduce a generic in-list type, which is largely opaque to kernel and which is always the RHS of a [NOT] IN expression. 

More and more, I wonder if 2/ is a better choice:
* Adding transparent array literals permanently and significantly widens the kernel API. Absent at least a second use case (preferably 3+ use cases), this seems like a lot of work and future liability for small gain?
* Forcing the engine to materialize the in-list as a rust array of (literal, type) pairs will blow up memory usage, and reduces the maximum in-list size an engine could otherwise handle. This is a "compile time" problem, independent of \what evaluation strategy the engine employs later.
* Using an opaque specific expression type for IN list allows the engine to do whatever it wants, without having to tell kernel _anything_. Kernel would just see a trait of some kind which it dutifully and blindly forwards to the engine's expression eval. We might even be able to generalize the "opaque engine expression" concept so that kernel can "handle" engine expression types it otherwise doesn't know about.

The downside of 2/ is it might make data skipping harder, since kernel doesn't have any idea what's "inside the box." It may be that we don't care, because long in-lists may turn out to not be good data skipping targets in the first place (low pruning power). If we did decide the pruning power is worth chasing, then we would need to expand the opaque in-list expression trait with method(s) that can perform data skipping. Probably very doable.

Reply:
I'd be okay with data skipping being a bit harder because another upside of 2 is that it allows us to generalize expressions in a way that allows the engine to decide what is correct. What I mean is currently there is no subset of expression functions that delta says an expression evaluator must have or what you can write in an expression. I've seen some very "interesting" expressions in the wild and making this opaque to the kernel save a trait to evaluate us would allow engines to define their own semantics here without worrying about whether it's a supported function or not. I know this doesn't have much to do with NOT/IN but just a point I figured was worth mentioning.

Reply:
Sounds good! Let's go with the opaque approach and worry about data skipping if/when needed

Reply:
If we _do_ eventually want to represent in-lists as opaque engine-defined expressions rather than array literals, is there a way to mark `ArrayData` as temporary/unstable/etc so we are able to remove it later without breaking the world? If nothing else, a doc comment explaining the situation might go a long way?